janusgraph-core/src/main/java/org/janusgraph/graphdb/query/graph/GraphCentricQueryBuilder.java

```diff
@@ -497,5 +407,122 @@ private static boolean coversAll(final MixedIndexType index, Condition<JanusGrap
         return match != null && indexInfo.supports(index, match, atom.getPredicate());
     }
 
+    private JointIndexQuery selectIndices(final Set<IndexType> indexCandidates, final MultiCondition<JanusGraphElement> conditions,
+                                          final Set<Condition> coveredClauses, final MutableBoolean isSorted) {
+        /* TODO: smarter optimization:
+        - use in-memory histograms to estimate selectivity of PredicateConditions and filter out low-selectivity ones
+                if they would result in an individual index call (better to filter afterwards in memory)
+        */
+        if (indexCandidates.size() <= tx.getGraph().getConfiguration().getIndexSelectThreshold()) {
+            return selectIndicesByBruteForce(indexCandidates, conditions, coveredClauses, isSorted);
+        } else {
+            return selectIndicesByApprox(indexCandidates, conditions, coveredClauses, isSorted);
+        }
```

Comment:
I would suggest to move `selectIndicesByBruteForce` and `selectIndicesByApprox` to separate classes. As for me they are strategies. So I would move them to `BruteForceIndexSelectionStrategy` and `ApproximationIndexSelectionStrategy`.

Reply:
I'm currently working on this approach but it looks like this requires moving more code than expected from the `GraphCentricQueryBuilder` to the strategy classes. I'll push a draft later today so you can at least have a look at what I mean.